src/core/job.c

```diff
@@ -1051,6 +1051,10 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
                         job_add_to_gc_queue(other->job);
                 }
 
+        /* Ensure that when an upheld unit activation job fails we requeue it, if it still needs to be
+         * uphold. If there are no state changes in the uphelder, it would not be retried otherwise. */
+        unit_submit_to_start_when_upheld_queue(u);
```

Comment:
as in the earlier comment: this should be inside the UNIT_FOREACH_DEPENDENCY() loop above, and operate on `other`, not on `u`, no?

Reply:
I don't think so? It works this way: A upholds B, B fails because of a dep, B's job here fails so B queues itself up again. I think we do want this.
We already have requeue on state changes of A, so it shouldn't be necessary to do it again here, no?

Reply:
waiting for a job is one thing. propagating failure from one job to another is another thing. I am pretty sure that we should recheck the Upholds= stuff also in the case where we delay one job for another, but do not propagate the failure.

I do think adding the unit itself to the queue probably also makes sense.

And actually I figure we probably should add the units to all three queues again.